test/functorch/test_control_flow.py

```diff
@@ -257,6 +257,1372 @@ def false_fn(x):
+        def forward(self, arg0_1: "f32[4]", arg1_1: "f32[4]", arg2_1: "f32[4, 4]", arg3_1: "f32[4]"):
+            # No stacktrace found for following nodes
+            view: "f32[1, 4]" = torch.ops.aten.view.default(arg3_1, [1, 4])
+            t: "f32[4, 4]" = torch.ops.aten.t.default(arg2_1);  arg2_1 = None
+            addmm: "f32[1, 4]" = torch.ops.aten.addmm.default(arg1_1, view, t);  arg1_1 = None
+            view_1: "f32[4]" = torch.ops.aten.view.default(addmm, [4]);  addmm = None
+            mul: "f32[4]" = torch.ops.aten.mul.Tensor(arg3_1, view_1)
+            mul_1: "f32[4]" = torch.ops.aten.mul.Tensor(arg0_1, arg3_1);  arg3_1 = None
+            mul_2: "f32[4]" = torch.ops.aten.mul.Tensor(arg0_1, view_1);  arg0_1 = view_1 = None
+            view_2: "f32[1, 4]" = torch.ops.aten.view.default(mul_1, [1, 4]);  mul_1 = None
+            t_1: "f32[4, 4]" = torch.ops.aten.t.default(t);  t = None
+            mm: "f32[1, 4]" = torch.ops.aten.mm.default(view_2, t_1);  t_1 = None
+            t_2: "f32[4, 1]" = torch.ops.aten.t.default(view_2)
+            mm_1: "f32[4, 4]" = torch.ops.aten.mm.default(t_2, view);  t_2 = view = None
+            t_3: "f32[4, 4]" = torch.ops.aten.t.default(mm_1);  mm_1 = None
+            sum_1: "f32[1, 4]" = torch.ops.aten.sum.dim_IntList(view_2, [0], True);  view_2 = None
+            view_3: "f32[4]" = torch.ops.aten.view.default(sum_1, [4]);  sum_1 = None
+            t_4: "f32[4, 4]" = torch.ops.aten.t.default(t_3);  t_3 = None
+            view_4: "f32[4]" = torch.ops.aten.view.default(mm, [4]);  mm = None
+            add: "f32[4]" = torch.ops.aten.add.Tensor(mul_2, view_4);  mul_2 = view_4 = None
+            return [view_3, t_4, add]""",  # noqa: B950
+        )
+        
+    def test_cond_in_forloop(self):
+        
+        def for_loop_fake(x):
+            for i in range(3):
+                x = x * x + 1
+            return x
+        
+        def for_loop_test(x):
+            for i in range(3):
+                pred = i < 3
+        
+                def true_fn(x):
+                    return x * x + 1
+
+                def false_fn(x):
+                    return x
+                
+                x = cond(pred, true_fn, false_fn, (x,))
+                
+            return x
+
+        x = torch.ones(4, requires_grad=True)
+        x_new = for_loop_test(x)
+        x_exp = for_loop_fake(x)
+        
+        self.assertEqual(x_new, x_exp)
+
+        grad_out = torch.ones_like(x_new)
+        grads = torch.autograd.grad(x_new, (x,), grad_out)
+        expected_grads = torch.autograd.grad(x_exp, (x,), grad_out)
+        self.assertEqual(expected_grads, grads)
+
+        def f(x):
+            x_new = for_loop_test(x)
+            grad_out = torch.ones_like(x_new)
+            return torch.autograd.grad(x_new, (x,), grad_out)
+
+        # TODO: During compilation, the metadata of the true_fn has the 
+        # requires_grad attribute set to False, while the false_fn has it
```

Comment:
Very good catch @ydwu4, I think this might even be the root cause of some of the other issues that we see. In this particular case, the `make_fx` is introducing this detach operation even in the forward path. So in case of
```
def forward(self, l_args_3_0_):
    l_args_3_0__1 = l_args_3_0_
    relu = torch.nn.functional.relu(l_args_3_0__1, inplace = False);  l_args_3_0__1 = None
    return (relu,)
```
`make_fx` turns this into
```
def forward(self, arg0_1):
    relu = torch.ops.aten.relu.default(arg0_1);  arg0_1 = None
    detach = torch.ops.aten.detach.default(relu)
    return (relu,)
```
I don't quite know what is the reason for this and I am not familiar with the inner workings of `make_fx`
This is the same for `torch.nn.GRUCell`. However, this does not happen if I just create a custom nn.Module that itself implements the ReLU function
```
class SimpleNN(torch.nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x):
        return torch.maximum(torch.zeros_like(x), x)
nn_module_false = SimpleNN()

def false_fn(x):
    return nn_module_false(x)
```
which when make_fx is applied gives
```
def forward(self, arg0_1):
    zeros_like = torch.ops.aten.zeros_like.default(arg0_1, pin_memory = False)
    maximum = torch.ops.aten.maximum.default(zeros_like, arg0_1);  zeros_like = arg0_1 = None
    return (maximum,)
```
Do you have any hints on how to approach this?

Reply:
yeah, I can take a look. For a normal function like below,
```python
import torch
from torch.fx.experimental.proxy_tensor import make_fx

def f(x):
    return torch.nn.functional.relu(x)

gm = make_fx(f)(torch.randn(3, 2))
```
we'll get a normal graph as shown below:
```python
class f(torch.nn.Module):
    def forward(self, x_1: "f32[3, 2]"):
        # No stacktrace found for following nodes
        relu: "f32[3, 2]" = torch.ops.aten.relu.default(x_1);  x_1 = None
        return relu
```
So the bug is probably related to cond.

Reply:
Thank you very much for looking into this!
I will continue to think about more such creative usecases, that potentially unveils bugs :-)

Some information that might help you in this issue: I tried to test several components of `torch.nn` and it seems that this issues is not present for all of them. For example, the detach operations `are not` present for 
- `torch.nn.Identity()`
- `torch.nn.Dropout()`
- `torch.nn.Linear()`
- `torch.nn.functional.threshold`
- `torch.nn.functional.glu`
- `torch.nn.functional.gelu`
- `torch.nn.functional.leaky_relu`
- `torch.nn.functional.logsigmoid`
- `torch.nn.functional.softplus`
- `torch.nn.functional.hardtanh`
- `torch.nn.functional.elu`
- `torch.nn.functional.selu`
- `torch.nn.functional.rrelu`

but, `are` present for
- `torch.nn.RNN`
- `torch.nn.LSTM`
- `torch.nn.GRU`
- `torch.nn.Transformer`
- `torch.nn.functional.relu`
- `torch.nn.functional.softmax`
- `torch.nn.functional.tanh`
- `torch.nn.functional.softmin`

At first, there was no clear pattern to me, but after quite some investigations into the failing cases, it seems that this behavior is mostly triggered by the `tanh`, `relu` or `softmax` functions. At least, all the failing cases I could track back to one of these functions. I hope this helps.

Reply:
Thanks for the thorough study!  @zou3519 told me when autograd saves a Tensor for backwards, there's a call to .detach() and pytorch has auto codegen for autograd: see `derivatives.yaml`. Generated code is in  `VariableTypeEverything.cpp`. In the case of `relu`, the result is saved like this:     `grad_fn->result_ = SavedVariable(result, true);`

Reply:
Ah, okay. Then this at least explains the behavior of what we see. To me it is interesting though why the `rrelu` or the `leaky_relu` functions do not show this behavir? Do these functions not save any tensor for the backward path?

Reply:
Good question. Not sure about the detail. Checked leaky_relu a bit: the difference between relu and leaky_relu in the generated code is that it doesn't store the SavedVariable in grad_fn like relu did. However, from derivatives.yarml, I cannot tell why we have such diference. They all return `auto_element_wise`. Might have other places of controlling the code-gen logic. This is a separate discussion though (or is this necessary for getting correct pytree output?).

Reply:
Okay, thank you.

I think for the moment, we could just leave a TODO, but we need to come back eventually to this, I guess. I also observed that a similar (potentially related) issue with the `requires_grad` flag occurs for the testcase [test_cond_in_forloop](https://github.com/pytorch/pytorch/blob/630549a7e12c32b0f0f9a366c53e247b9a32a230/test/functorch/test_control_flow.py#L678), for example. In that case, there is no `relu` function involved, but still the `true_out.meta["tensor_meta"]` has `requires_grad=False`, while `true_out.meta["tensor_meta"]` has set it to True

Reply:
Can you check how `tensor_meta` is created and why it is set to False in one branch?

Reply:
I tried to go down the rabbit hole and actually found something puzzling to me. In the simple case, 
```
def true_fn(x):
    return x.sin()
def false_fn(x):
    return x.cos()
```
The metadata are
```
true_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
false_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
```
However, if in one branch the input arguments are just passed through, e.g.,
```
def true_fn(x):
    return x
def false_fn(x):
    return x.cos()
```
the metadata become
```
true_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=True, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
false_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
```

If one applies even a simple multiplication with 1 to the input argument, the metadata match again
```
def true_fn(x):
    return x*1
def false_fn(x):
    return x.cos()
true_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
false_out.meta["tensor_meta"] = TensorMetadata(shape=torch.Size([4]), dtype=torch.float32, requires_grad=False, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}))
```
To me this behavior is unexpected, but I don't know where this comes from.

Reply:
@ydwu4. Do you maybe have any insights into this? Thank you